src/main/java/com/github/packageurl/PackageURL.java

```
@@ -347,22 +344,31 @@ private String validatePath(final String value, final boolean isSubpath) throws
     }
 
-    private String validatePath(final String[] segments, final boolean isSubpath) throws MalformedPackageURLException {
+    private static boolean isValidSegment(String segment, final boolean isSubpath) {
+        return (!isSubpath || (!segment.isEmpty() && !".".equals(segment) && !"..".equals(segment)));
+    }
+
+    private static String validatePath(final String[] segments, final boolean isSubpath) throws MalformedPackageURLException {
         if (segments == null || segments.length == 0) {
             return null;
         }
+
         try {
             return Arrays.stream(segments)
                     .map(segment -> {
-                        if (isSubpath && ("..".equals(segment) || ".".equals(segment))) {
-                            throw new ValidationException("Segments in the subpath may not be a period ('.') or repeated period ('..')");
-                        } else if (segment.contains("/")) {
-                            throw new ValidationException("Segments in the namespace and subpath may not contain a forward slash ('/')");
-                        } else if (segment.isEmpty()) {
-                            throw new ValidationException("Segments in the namespace and subpath may not be empty");
+                        if (!isSubpath) {
+                            if ("..".equals(segment) || ".".equals(segment)) {
+                                throw new ValidationException("Segments in the namespace may not be a period ('.') or repeated period ('..')");
+                            } else if (segment.contains("/")) {
+                                throw new ValidationException("Segments in the namespace and subpath may not contain a forward slash ('/')");
+                            } else if (segment.isEmpty()) {
+                                throw new ValidationException("Segments in the namespace and subpath may not be empty");
+                            }
                         }
```

The previous condition was better.

According to the spec a decoded `namespace` segment:

> When percent-decoded, a segment:
> - must not contain a '/'
> - must not be empty

For a `subpath` segment there is an additional limitation regarding `.` and `..`.

---

The previous code:

1. did not allow  `"."` and `".."` in `subpath`
2. did not allow `'/'` or `""` in either `subpath` or `namespace`

Looking at it now, I can't remember what happens to the first condition in `namespace`, but the first condition is _legal_ in the `subpath`. This is the reason for the change. I agree I have the same confusion with the wording of the spec. Please read the issue I opened there that explains the reasoning behind it and then let me know if there's still an issue.

---

The way I see it, there are two parts in the specification:

- The specification of a **lenient** parser (see [How to parse a `purl` string in its components](https://github.com/package-url/purl-spec/blob/main/PURL-SPECIFICATION.rst#how-to-parse-a-purl-string-in-its-components)), which allows for fixes, such as dropping `.`, `..` or empty segments. This is done in:

   https://github.com/package-url/packageurl-java/blob/ac62676e41ff7c11a2557db104d5d1cce3993c1a/src/main/java/com/github/packageurl/PackageURL.java#L638

- The specification of the components of a `purl` (see [Rules for each `purl` component](https://github.com/package-url/purl-spec/blob/main/PURL-SPECIFICATION.rst#rules-for-each-purl-component), which does **not** allow for `.`, `..` or empty segments and segments containing `/`. The condition above is part of:

   https://github.com/package-url/packageurl-java/blob/ac62676e41ff7c11a2557db104d5d1cce3993c1a/src/main/java/com/github/packageurl/PackageURL.java#L351

   which is called **after** `parsePath`, so both parts of the specification are respected.

**Note**: if the user calls the constructor with multiple arguments, invalid segments are not discarded, but that is probably correct, since that constructor is not lenient.

---

I am complaining over on the spec that it's not clear _to me, at least_, that one section refers to the parser, and one referring to the component. An example I gave is if you use language like _MUST NOT_, then how is it possible that it exists later to silently discard? In that case, it should use _MAY_, or clearly state that it's talking about two entirely different things, which I do not think it does.

---

There's an open discussion about this below, so I am resolving this one.